modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java

```diff
@@ -459,17 +461,42 @@ private long absFileIdx(CheckpointEntry pointer) {
      * @param partsCounter Partition mapped to update counter.
      * @return Earliest WAL pointer for group specified.
      */
-    @Nullable public WALPointer searchEarliestWalPointer(int grpId, Map<Integer, Long> partsCounter) throws IgniteCheckedException {
+    @Nullable public WALPointer searchEarliestWalPointer(
+        int grpId,
+        Map<Integer, Long> partsCounter,
+        long margin
+    ) throws IgniteCheckedException {
         if (F.isEmpty(partsCounter))
             return null;
 
         Map<Integer, Long> modifiedPartsCounter = new HashMap<>(partsCounter);
 
         FileWALPointer minPtr = null;
 
+        LinkedList<T3<Map.Entry<Integer, Long>, FileWALPointer, Long>> historyPointerCandidat = new LinkedList<>();
```

Comment:
Let's create a class T3<Map.Entry<Integer, Long>, FileWALPointer, Long> with a proper names for fields.
Code looks deliberately obfuscated.

Reply:
Probably a typo: candidat**e**